File: Pharmaceutical_Company_Example/Structural/Python/Decorator.py
# Base class
class RegulatoryProcess:
    def execute(self):
        raise NotImplementedError("This method should be implemented by subclasses.")


# Concrete class
class BasicRegulatoryStep(RegulatoryProcess):
    def __init__(self, step_name):
        self.step_name = step_name

    def execute(self):
        print(f"Executing step: {self.step_name}")


# Decorator Base Class for adding additional features to RegulatoryProcess
class ProcessDecorator(RegulatoryProcess):
    def __init__(self, decorated_process):
        self.decorated_process = decorated_process

    def execute(self):
        self.decorated_process.execute()


# Concrete Decorator - Adds Auditing feature to the regulatory process
class AuditingDecorator(ProcessDecorator):
    def execute(self):
        super().execute()
        self.add_audit()

    def add_audit(self):
        print("Adding auditing to step.")


# Concrete Decorator - Adds Documentation requirement to the process
class DocumentationDecorator(ProcessDecorator):
    def execute(self):
        super().execute()
        self.add_documentation()

    def add_documentation(self):
        print("Ensuring documentation is complete for step.")


# Concrete steps
clinical_trial_step = BasicRegulatoryStep("Clinical Trial Approval")
safety_review_step = BasicRegulatoryStep("Safety Review")

# Decorate the clinical trial step with auditing and documentation features
clinical_trial_with_audit = AuditingDecorator(clinical_trial_step)
clinical_trial_with_full_features = DocumentationDecorator(clinical_trial_with_audit)

# Decorate the safety review step with documentation only
safety_review_with_documentation = DocumentationDecorator(safety_review_step)

# Execute both processes
print("Executing Clinical Trial Step with Audit and Documentation:")
clinical_trial_with_full_features.execute()

print("\nExecuting Safety Review Step with Documentation:")
safety_review_with_documentation.execute()

File: Pharmaceutical_Company_Example/Structural/JavaScript/Decorator.js
// Base class (Abstract class)
class RegulatoryProcess {
    execute() {
        throw new Error("This method should be implemented by subclasses.");
    }
}

// Concrete class - BasicRegulatoryStep
class BasicRegulatoryStep extends RegulatoryProcess {
    constructor(stepName) {
        super();
        this.stepName = stepName;
    }

    execute() {
        console.log(`Executing step: ${this.stepName}`);
    }
}

// Decorator Base Class for adding additional features to RegulatoryProcess
class ProcessDecorator extends RegulatoryProcess {
    constructor(decoratedProcess) {
        super();
        this.decoratedProcess = decoratedProcess;
    }

    execute() {
        this.decoratedProcess.execute();
    }
}

// Concrete Decorator - Adds Auditing feature to the regulatory process
class AuditingDecorator extends ProcessDecorator {
    execute() {
        super.execute();
        this.addAudit();
    }

    addAudit() {
        console.log("Adding auditing to step.");
    }
}

// Concrete Decorator - Adds Documentation requirement to the process
class DocumentationDecorator extends ProcessDecorator {
    execute() {
        super.execute();
        this.addDocumentation();
    }

    addDocumentation() {
        console.log("Ensuring documentation is complete for step.");
    }
}

// Concrete steps
const clinicalTrialStep = new BasicRegulatoryStep("Clinical Trial Approval");
const safetyReviewStep = new BasicRegulatoryStep("Safety Review");

// Decorate the clinical trial step with auditing and documentation features
const clinicalTrialWithAudit = new AuditingDecorator(clinicalTrialStep);
const clinicalTrialWithFullFeatures = new DocumentationDecorator(clinicalTrialWithAudit);

// Decorate the safety review step with documentation only
const safetyReviewWithDocumentation = new DocumentationDecorator(safetyReviewStep);

// Execute both processes
console.log("Executing Clinical Trial Step with Audit and Documentation:");
clinicalTrialWithFullFeatures.execute();

console.log("\nExecuting Safety Review Step with Documentation:");
safetyReviewWithDocumentation.execute();
